app-tracking-protection/vpn-impl/src/main/java/com/duckduckgo/mobile/android/vpn/service/state/RealVpnStateMonitor.kt

```diff
@@ -74,6 +76,30 @@ class RealVpnStateMonitor @Inject constructor(
             .distinctUntilChanged()
     }
 
+    override suspend fun isAlwaysOnEnabled(): Boolean {
+        return vpnServiceStateStatsDao.getLastStateStats()?.alwaysOnState?.alwaysOnEnabled ?: false
+    }
+
+    override suspend fun vpnLastDisabledByAndroid(): Boolean {
+        fun vpnUnexpectedlyDisabled(): Boolean {
+            return vpnServiceStateStatsDao.getLastStateStats()?.let {
+                (
+                    it.state == DISABLED &&
+                        it.stopReason != SELF_STOP &&
+                        it.stopReason != REVOKED
+                    )
+            } ?: false
+        }
+
+        fun vpnKilledBySystem(): Boolean {
+            val lastHeartBeat = vpnHeartBeatDao.hearBeats().maxByOrNull { it.timestamp }
+            return lastHeartBeat?.type == VpnServiceHeartbeatMonitor.DATA_HEART_BEAT_TYPE_ALIVE &&
+                !vpnFeaturesRegistry.isAnyFeatureRegistered()
```

Comment:
This used to be !vpnFeaturesRegistry.isFeatureRegistered(AppTpVpnFeature.APPTP_VPN). I think this check will scale better.